bundles/org.openhab.ui.start/src/main/java/org/openhab/ui/start/internal/RootServlet.java

```diff
@@ -0,0 +1,172 @@
+        STARTING,
+        STARTED,
+        STOPPING,
+        UPDATING;
+    }
+
+    private String page404;
+    private String pageStatus;
+
+    private DashboardReady dashboardStarted;
+    private LifeCycleState lifecycleState = LifeCycleState.STARTING;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        if (dashboardStarted != null) {
+            // all is up and running
+            if (req.getRequestURI().equals("/")) {
+                resp.sendRedirect("/start/index");
+            } else {
+                resp.sendError(HttpServletResponse.SC_NOT_FOUND);
+                resp.setContentType("text/html;charset=UTF-8");
+                resp.getWriter().append(page404);
+                resp.getWriter().close();
+            }
+        } else {
+            // report current system state
+            String message = null;
+            String subMessage = null;
+            switch (lifecycleState) {
+                case STARTING:
+                    message = "openHAB is starting...";
+                    subMessage = "Please wait a moment!";
+                    break;
+                case UPDATING:
+                    message = "openHAB is updating...";
+                    subMessage = "Please wait a moment!";
+                    break;
+                case STOPPING:
+                    message = "openHAB is shutting down...";
+                    subMessage = "Please stand by.";
+                    break;
+                default:
+                    throw new IllegalStateException("Invalid system state " + lifecycleState);
+            }
+            resp.setContentType("text/html;charset=UTF-8");
+            resp.getWriter().append(pageStatus.replace("${message}", message).replace("${submessage}", subMessage));
+            resp.getWriter().close();
+        }
+    }
+
+    @Activate
+    protected void activate(ComponentContext context) {
+        try {
+            httpService.registerServlet("/", this, new Properties(), httpService.createDefaultHttpContext());
+        } catch (ServletException | NamespaceException e) {
+            logger.error("Failed registering root servlet!", e);
+        }
+        URL notfound = context.getBundleContext().getBundle().getEntry("pages/404.html");
+        if (notfound != null) {
+            try {
+                page404 = IOUtils.toString(notfound.openStream());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            throw new RuntimeException("Cannot find 404.html - failed to initialize root servlet");
+        }
+        URL status = context.getBundleContext().getBundle().getEntry("pages/status.html");
+        if (status != null) {
+            try {
+                pageStatus = IOUtils.toString(status.openStream());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
Imho, the code could benefit a lot from a better separation between the error handling and the logic, what about not catching this error and marking the method to throw an IOException? If that is allowed otherwise splitting of some methods could be beneficial?

Reply:
This is the activation of the component, so we cannot declare catched exceptions.
And these errors really mean that the component isn't usable, so I think this should happen in here.

Reply:
Maybe splitting up the method could help but, but if at least the exceptions are cleaned 👍